### PR TITLE
Sp gtfs

### PR DESCRIPTION
**Pull request by carlhiggs**
I believe this could be good to go, pending checks by others --- there's a slight change,

You need to download the GTFS geopackage (dated 27 august), and the gtfs_config.py and locate these in data/GTFS/...

Then, if you run "python setup_config.py" this generates config files (for sample points, these are python scripts now - hope that's okay) --- and then you could use 
bash ./process_regions.sh city city city 
to iterate over your cities
I haven't checked the gpkg outputs yet -- but it seems to all work (following interactive checks); it runs successfully.  I'll check outputs tomorrow for my cities.  Fingers crossed, all good !

### CODE REVIEW

**Comment by carlhiggs** [comment]
Just an update --- had a quick look at walkability for Auckland --- looks plausible.  I'll check out the outputs tomorrow properly, but for now - it looks like its run, and is quick to run.  I've uploaded my cities to the **Global Indicators - data/output/September 2020** folder on CloudStor.   @shiqin-liu  @gboeing @duelran

**Comment by carlhiggs** [comment]
Hi @shiqin-liu  and @gboeing  --- I just made a commit updating th agg.py and setup scripts to work with the new variables when outputing to city and hex aggregations.   I've run this with one city (Auckland) as a minimal test and can confirm that it now outputs the hex250, and all cities geopackages, with output apparently as intended.  The study region name in both hex and all cities layers is now title case with space replacing underscore, which might help with plotting; and I also explicitly ordered the hex output columns.  Otherwise - the code should be basically as before.    (these commits won't impact results of analyses since the bug fixes I updated and advised on earlier today - so no need to re-run sample results for this)

File: process/setup_sp.py [hunk 1]
```diff
@@ -407,44 +401,31 @@ def create_full_nodes(
...
-    # Define a lambda function to compute the weighted mean:
-    # wm = lambda x: np.average(x, weights=full_nodes.loc[x.index, "density_weight"])
-    # join up full nodes with density fields
-    density_fields = [pop_density, intersection_density]
-    full_nodes = full_nodes.join(gdf_nodes_simple[density_fields], on="node", how="left")
+    for d in distance_names:
+        full_nodes[d] = full_nodes[d] + full_nodes["node_distance_m"]
+        distance_fields.append(d)
+    distance_names = [x for x in distance_names if x in gdf_nodes_poi_dist.columns]
+    print("\t - calculate proximity-weighted average of densitiy statistics for each sample point")
```
Comment (gboeing) [inline process/setup_sp.py]: minor typo: "densitiy"

**Review by gboeing — COMMENTED** [review]
@carlhiggs I did a visual perusal of your PR this morning. It generally looks fine to me. I haven't got time (or spare CPUs) to run it for a few days... perhaps @shiqin-liu can do a quick run before merging this?

**Comment by shiqin-liu** [comment]
@gboeing sure, I try to test run city of Vic from this branch, what I did is:

- Download GTFS .gpkg and gtfs_config from Cloudstor in the data/GTFS folder
- Re-download the input .gpkg from Cloudstor input folder
- Run python setup_config.py
- Run python sp.py vic

But I got an error:
```
(base) root@b27971c88234:/home/jovyan/work/process# python sp.py vic

Global indicators project 2020-09-04

Process city: Vic

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.000550469 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 8 threads.
Setting CH node vector of size 4844
Setting CH edge vector of size 14080
Range graph removed 14202 edges of 28160
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
Create hex_id for sample points
Creating long form working dataset of sample points to evaluate respective node distances and densities
	 - create long form dataset
	 - join POIs results from nodes to sample points
	 - calculate proximity-weighted average of densitiy statistics for each sample point
Calculating sample point specific analyses ...
	 - Best PT (any) access score
Traceback (most recent call last):
  File "sp.py", line 289, in <module>
    samplePointsData[var] = samplePointsData[columns].max(axis=axis)
  File "/opt/conda/lib/python3.7/site-packages/geopandas/geodataframe.py", line 576, in __getitem__
    result = super(GeoDataFrame, self).__getitem__(key)
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py", line 2806, in __getitem__
    indexer = self.loc._get_listlike_indexer(key, axis=1, raise_missing=True)[1]
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/indexing.py", line 1553, in _get_listlike_indexer
    keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/indexing.py", line 1646, in _validate_read_indexer
    raise KeyError(f"{not_found} not in index")
KeyError: "['sp_access_pt_gtfs_any_binary'] not in index"
```


I have not figured out why this occurs... @carlhiggs could you help me take a look if you get a chance? Thanks in advance!

**Comment by carlhiggs** [comment]
Hi,

Thanks for having a look at things. That's curious about the error with Vic
--- perhaps it may arise in the case where a city didn't have a gtfs feed
defined (Vic doesn't); in theory it should be fine though and just yield a
null.  I ran it for six cities without issue - but I'll see if I can get a
chance to check out Vic later today.

Regarding time, it doesn't take days to calculate access - should take
under 10 mins for most cities (Auckland was under 6 from memory, I didn't
pay close attention to others - but the longer processing was taken in the
last round with local neighbourhood statistics --- assuming we all left
things running last week?).

In any case - we have to run it at some point to finish our analyses really
... And the sooner we attempt to run it for all cities, if there are
exceptions that arise, the sooner we'll find them.

**Comment by gboeing** [comment]
> In any case - we have to run it at some point to finish our analyses really ... And the sooner we attempt to run it for all cities, if there are exceptions that arise, the sooner we'll find them.

@carlhiggs Agreed! I misspoke earlier, I've got 6 CPUs available: it's my RAM that's maxed out (95% r.n.) with the current process runs. It should be finished this weekend but in the meantime I'm risk-averse with loading anything else at the same time that could wipe me out (and lose all my current runtime progress). I can run this when that's finished, time permitting (got full time childcare duties the next 2 days)...

**Comment by shiqin-liu** [comment]
Sorry, I accidentally close the PR.. so reopen again

@carlhiggs I try to run Olomouc, and the same error occur, maybe it is because of the null GTFS?

I run Odense, there is another error:
```
(base) root@b27971c88234:/home/jovyan/work/process# python sp.py odense

Global indicators project 2020-09-05

Process city: Odense

Reproject network, and save the projected network to disk
Create study region sample point output file
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.000598888 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 8 threads.
Setting CH node vector of size 29523
Setting CH edge vector of size 78638
Range graph removed 80002 edges of 157276
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Traceback (most recent call last):
  File "sp.py", line 234, in <module>
    output_prefix = 'sp_nearest_node_')
  File "/home/jovyan/work/process/setup_sp.py", line 356, in cal_dist_node_to_nearest_pois
    gdf_poi_filtered["y"],
  File "/opt/conda/lib/python3.7/site-packages/pandana/network.py", line 551, in set_pois
    node_ids = self.get_node_ids(x_col, y_col)
  File "/opt/conda/lib/python3.7/site-packages/pandana/network.py", line 388, in get_node_ids
    distances, indexes = self.kdtree.query(xys.values)
  File "sklearn/neighbors/_binary_tree.pxi", line 1331, in sklearn.neighbors._kd_tree.BinaryTree.query
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 73, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 646, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 100, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

I finished my cities and uploaded them to the output folder. Carl, could you try these cities, if the issue does not occur at your end, I will re-check if I have missed anything...

**Comment by carlhiggs** [comment]
Hi @gboeing and @shiqin-liu --- thanks again for raising this issue, now fixed.  For the case where a city doesn't have an 'nearest node' indicator, when the null variable was created and named, i missed adding the 'sp_nearest_node_' prefix.  I added that in and all good --- Vic processes in 20 seconds successfully, apparently:
image

Fair point regarding the processes Geoff ---  given its the weekend (soon for you I think!) no worries leaving this until next week anyway; I expect your cities should be complete in an hour or so then, and I completely understand not wanting to compromise what you're otherwise doing; all good.

That amendment has now been pushed, and I expect should deal with the Olomouc issue.

Catch you all next week, Carl

**Comment by carlhiggs** [comment]
Thanks Shirley and Geoff - I only saw your comment re: uploading cities
after i posted my update above; i just downloaded all the cities that were
uploaded and will leave these running --- fingers crossed all good, and
we'll see where we're at on Monday!

Have a great weekend,

carl

**Comment by gboeing** [comment]
> and then you could use bash ./process_regions.sh city city city to iterate over your cities

I don't see any script called `process_regions.sh` in this branch.

**Comment by gboeing** [comment]
I ran it all manually one city at a time. Appears to work. I append my terminal output below.

I don't see documentation for the changes made in this branch... we'll want to make sure all documentation and instructions are updated prior to merging into master. @nicholas-david can you coordinate with @carlhiggs to add/update the docs for new functionality and usage this coming week?

In the meantime, I'm unclear where the results are being saved and if we should be (re-)uploading data outputs at this point after running the process in this branch? I'll check back Monday if action items remain.

```
(base) root@0004c2bb7f95:/home/jovyan/work# python setup_config.py 
Generate study region configuration json file
Generate cities aggregation configuration json file
All study region configuration file were generated, total time is : 0.00 hours or 0.00 seconds
(base) root@0004c2bb7f95:/home/jovyan/work# python sp.py phoenix true

Global indicators project 2020-09-05

Process city: Phoenix

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.00121052 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 12 threads.
Setting CH node vector of size 172056
Setting CH edge vector of size 492562
Range graph removed 506940 edges of 985124
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Create hex_id for sample points
Creating long form working dataset of sample points to evaluate respective node distances and densities
	 - create long form dataset
	 - join POIs results from nodes to sample points
	 - calculate proximity-weighted average of densitiy statistics for each sample point
Calculating sample point specific analyses ...
	 - Best PT (any) access score
	 - Daily living score
	 - Walkability index
Save to geopackage...
Total time is : 5.12 minutes
(base) root@0004c2bb7f95:/home/jovyan/work# python sp.py seattle true

Global indicators project 2020-09-05

Process city: Seattle

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.00243959 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 12 threads.
Setting CH node vector of size 359111
Setting CH edge vector of size 986596
Range graph removed 1010544 edges of 1973192
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Create hex_id for sample points
Creating long form working dataset of sample points to evaluate respective node distances and densities
	 - create long form dataset
	 - join POIs results from nodes to sample points
	 - calculate proximity-weighted average of densitiy statistics for each sample point
Calculating sample point specific analyses ...
	 - Best PT (any) access score
	 - Daily living score
	 - Walkability index
Save to geopackage...
Total time is : 11.37 minutes
(base) root@0004c2bb7f95:/home/jovyan/work# python sp.py mexico_city true

Global indicators project 2020-09-05

Process city: Mexico_City

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.00295682 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 12 threads.
Setting CH node vector of size 447873
Setting CH edge vector of size 1225468
Range graph removed 1234136 edges of 2450936
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Create hex_id for sample points
Creating long form working dataset of sample points to evaluate respective node distances and densities
	 - create long form dataset
	 - join POIs results from nodes to sample points
	 - calculate proximity-weighted average of densitiy statistics for each sample point
Calculating sample point specific analyses ...
	 - Best PT (any) access score
	 - Daily living score
	 - Walkability index
Save to geopackage...
Total time is : 15.23 minutes
(base) root@0004c2bb7f95:/home/jovyan/work# python sp.py melbourne true

Global indicators project 2020-09-05

Process city: Melbourne

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.0022413 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 12 threads.
Setting CH node vector of size 375303
Setting CH edge vector of size 1070182
Range graph removed 1082596 edges of 2140364
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Create hex_id for sample points
Creating long form working dataset of sample points to evaluate respective node distances and densities
	 - create long form dataset
	 - join POIs results from nodes to sample points
	 - calculate proximity-weighted average of densitiy statistics for each sample point
Calculating sample point specific analyses ...
	 - Best PT (any) access score
	 - Daily living score
	 - Walkability index
Save to geopackage...
Total time is : 10.68 minutes
(base) root@0004c2bb7f95:/home/jovyan/work# python sp.py sydney true

Global indicators project 2020-09-05

Process city: Sydney

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.00148019 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 12 threads.
Setting CH node vector of size 253304
Setting CH edge vector of size 703742
Range graph removed 715770 edges of 1407484
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Create hex_id for sample points
Creating long form working dataset of sample points to evaluate respective node distances and densities
	 - create long form dataset
	 - join POIs results from nodes to sample points
	 - calculate proximity-weighted average of densitiy statistics for each sample point
Calculating sample point specific analyses ...
	 - Best PT (any) access score
	 - Daily living score
	 - Walkability index
Save to geopackage...
Total time is : 7.51 minutes
(base) root@0004c2bb7f95:/home/jovyan/work# 
```

**Review by shiqin-liu — COMMENTED** [review]
@carlhiggs a quick update, I ran my cities with `bash ./process_regions.sh vic olomouc lisbon valencia odense graz ghent belfast bern chennai barcelona`

It took less than an hour, All cities work except Odense, I still got this error:

```
Global indicators project 2020-09-06

Process city: Odense

Read network from disk.
Study region sample point output file exists
Read poplulation and intersection density from local file.
Time taken to calculate or load city local neighbourhood statistics: 0.000558356 mins

Calculate assessbility to POIs.
Generating contraction hierarchies with 8 threads.
Setting CH node vector of size 29523
Setting CH edge vector of size 78638
Range graph removed 80002 edges of 157276
. 10% . 20% . 30% . 40% . 50% . 60% . 70% . 80% . 90% . 100%

Calculating nearest node analyses ...

	- Open street map destinations
		['fresh_food_market', 'convenience', 'pt_osm_any']

	- Public open space
		['public_open_space_any']
		['public_open_space_large']

	- Public transport (GTFS)
		['pt_gtfs_any', 'pt_gtfs_freq_30', 'pt_gtfs_freq_20']
Traceback (most recent call last):
  File "sp.py", line 234, in <module>
    output_prefix = 'sp_nearest_node_')
  File "/home/jovyan/work/process/setup_sp.py", line 356, in cal_dist_node_to_nearest_pois
    gdf_poi_filtered["y"],
  File "/opt/conda/lib/python3.7/site-packages/pandana/network.py", line 551, in set_pois
    node_ids = self.get_node_ids(x_col, y_col)
  File "/opt/conda/lib/python3.7/site-packages/pandana/network.py", line 388, in get_node_ids
    distances, indexes = self.kdtree.query(xys.values)
  File "sklearn/neighbors/_binary_tree.pxi", line 1331, in sklearn.neighbors._kd_tree.BinaryTree.query
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 73, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 646, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 100, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

```

Is this error occur in yours?

**Comment by carlhiggs** [comment]
Hi Shirley and Geoff -- thanks for running the code also, and for picking up on this error.  This will relate to some bug (yet to fix) in the GTFS analysis code; for some reason, the Odense GTFS stop layer does not have geometries it seems -- and a lot of Nulls --- for example, in the below screenshot, the upper table shoes Odense GTFS stops (nulls for lat and lon, amongst other fields), and the lower shows a more usual output which does have the coordinates required to generate geometries...... so, I'll go back and look at that code which will have given rise to this.  In short, this is a bug outside this code -- however, ideally a more informative error message could be thrown by this code in future (e.g. if no valid geometries were found in input data) which could lead us to find the root cause more rapidly.    I'll try to have a look at this today --- but it seems good news that the process has worked for most other cities 

image

**Comment by carlhiggs** [comment]
I believe I have found the issue, and its to do with the data not the code per se (although, both code bases could raise better errors, perhaps).  What has happened is, the stop_id in the Danish GTFS stop_times and stops files don't match --- both are interpreted as strings when parsed by UrbanAccess (which is fine, and appropriate), however, because the stops.txt stop_id contains a series of zeroes as a prefix, when the data which has been analysed for headways is to be joined with locations defined stops.txt no matches are found --- as per image below.  I think the fix to this is manual cleaning of stops.txt to remove the prefix --- or alternately, perhaps they could both be treated as integers which would have the same effect (but that isn't a reliable operation to use on GTFS feeds which may have non-numeric IDs).  I'll update later once I've addressed this, and re-upload the GTFS gpkg
image

**Comment by carlhiggs** [comment]
I have now created a fix to the Odense GTFS input data (saving a copy of the original, with a note explaining this manual cleaning) and this should resolve the Odense issue described above.  The GTFS stops for Odense now look like this, which looks right (previously, it lacked geometry due to the mismatch on id described above -- which was of course wrong):
image

The note I left on the fix, in a file names 'stops_cleaned_readme.txt'  was this:

> Stops.txt was modified to fix an issue with zero prefixes to IDs which were not present in stop_times.txt and which prevented match up on ID to join locations to stop_times. 
> 
> To address this issue, it was first verified that linkage based on interpreting stop_id as a string representation of an integer (which stripped zeroes) worked successfully.  The actual file was modified by opening as a csv file in Microsoft Excel (which implicitly stripped the zeroes from the integers), and then saving, closing, and renaming to .txt.  A copy of the original stops.txt was saved as stops_original.txt.
> 
> For example
> 
> 000008600512 became 8600512, so that it could be joined with stop_id 8600512 in stops_times.txt

Revised data (gpkg and feed) is up on CloudStor, and this appears to work now:
image

**Comment by carlhiggs** [comment]
Just confirming that I believe this branch is appropriate for merging into the master --- it contains all the changes that the latest all cities comparisons were run with

File: process/setup_sp.py (hunk 1 above)
Comment (gboeing) [reply process/setup_sp.py]: @carlhiggs could you fix this one typo here?

Comment (gboeing) [reply process/setup_sp.py]: NVM, I fixed it directly.

**Pull request merged**